specs/source.go

```diff
@@ -21,13 +21,29 @@ type SourceSpec struct {
 }
 
 func (s *SourceSpec) UnmarshalYAML(n *yaml.Node) error {
-	if err := n.Decode(s); err != nil {
+	var sourceSpecWithoutUnmarshal struct {
```

Comment:
This can be some other way, such as adding an inner struct inside SourceSpec etc. but handling it in other places gets tedious, and we shouldn't leak implementation details into public APIs.

Reply:
Why we need to duplicate the spec definition here?

Reply:
because `yaml.Decode` calls `UnmarshalYAML` (if there is one) which then calls `Decode` and so on.

Reply:
Fixed by adding a new method, `Validate() error` which is called after unmarshal.